engine/telemetry/pubsub.go

```diff
@@ -0,0 +1,826 @@
+		clients:    map[string]*activeClient{},
+
+		spanClients: map[spanKey]string{},
+		spanParents: map[spanKey]trace.SpanID{},
+		spanDone:    map[spanKey]bool{},
+	}
+	mux.HandleFunc("/v1/traces", ps.TracesHandler)
+	mux.HandleFunc("/v1/logs", ps.LogsHandler)
+	mux.HandleFunc("/v1/metrics", ps.MetricsHandler)
+	return ps
+}
+
+// Processor returns a span processor that keeps track of client IDs,
+// inheriting them from parent spans if needed.
+func (ps *PubSub) Processor() sdktrace.SpanProcessor {
+	return clientTracker{ps}
+}
+
+type clientTracker struct{ *PubSub }
+
+// OnStart sets the client ID on the span, if it is not already present,
+// inheriting it from its parent span if needed by keeping track of all span
+// clients.
+func (p clientTracker) OnStart(ctx context.Context, span sdktrace.ReadWriteSpan) {
+	p.spansL.Lock()
+	defer p.spansL.Unlock()
+	clientID := p.extractClient(ctx, span)
+	if clientID != "" {
+		span.SetAttributes(attribute.String(telemetry.ClientIDAttr, clientID))
+	}
+	p.trackSpan(span)
+}
+
+// OnEnd does nothing. Span client state must be cleaned up when the client
+// goes away, not when a span completes.
+func (p clientTracker) OnEnd(span sdktrace.ReadOnlySpan) {
+	p.spansL.Lock()
+	p.spanDone[spanKey{
+		TraceID: span.SpanContext().TraceID(),
+		SpanID:  span.SpanContext().SpanID(),
+	}] = true
+	p.spansL.Unlock()
+}
+
+// Shutdown does nothing.
+func (clientTracker) Shutdown(context.Context) error { return nil }
+
+// ForceFlush does nothing.
+func (clientTracker) ForceFlush(context.Context) error { return nil }
+
+func (p clientTracker) extractClient(ctx context.Context, span sdktrace.ReadOnlySpan) string {
+	// slog := slog.With("span", span.Name(), "spanID", span.SpanContext().SpanID())
+
+	// slog.ExtraDebug("!!! ClientAnnotator tracking span clients")
+
+	for _, attr := range span.Attributes() {
+		if attr.Key == telemetry.ClientIDAttr {
+			// slog.ExtraDebug("!!! > found existing client ID", "value", attr.Value.AsString())
+			return attr.Value.AsString()
+		}
+	}
+
+	var clientID string
+	metadata, err := ClientMetadataFromContext(ctx)
+	if err == nil {
+		clientID = metadata.ClientID
+		// slog.ExtraDebug("!!! > adding client ID from ctx", "client", clientID)
+	} else if span.Parent().IsValid() {
+		// NOTE: empty result if not found is OK
+		clientID = p.spanClients[spanKey{
+			TraceID: span.SpanContext().TraceID(),
+			SpanID:  span.Parent().SpanID(),
+		}]
+		// slog.ExtraDebug("!!! > inheriting client ID from parent", "client", clientID)
+	}
+
+	return clientID
+}
+
+// clientsFor returns the relevant client IDs for a given span, traversing
+// through parents, in random order.
+func (ps *PubSub) clientsFor(traceID trace.TraceID, spanID trace.SpanID) []string {
+	ps.spansL.Lock()
+	defer ps.spansL.Unlock()
+	key := spanKey{
+		TraceID: traceID,
+		SpanID:  spanID,
+	}
+	// slog := slog.With("traceID", traceID, "spanID", spanID)
+	clients := map[string]struct{}{}
+	for {
```

Comment:
nit: in the interest of paranoia, is it worth some sanity check for the conditions that would result in this loop accidentally spinning forever? e.g. checking `parent` is not the same value as the existing `SpanID`. That would require a bug elsewhere but would rather error if that happens than get stuck in a loop.

Haven't fully loaded all context yet so may be a dumb suggestion, not a blocker.

Reply:
Yeah sure, might as well. https://github.com/dagger/dagger/pull/7452/commits/1515f71e2b695264d27407cf28732d633a709473 makes it keep track of which keys it's seen as it goes and bail if it sees one again. That should also detect cycles.